Add singleTransaction option for ORMExecutor (#388)
doctrine/data-fixtures: docs/en/how-to/loading-fixtures.rst

    disable it.

If you want to append the fixtures instead of purging before loading
then pass ``append: true`` to the ``execute()`` method:

.. code-block:: php
    <?php
    $executor->execute($loader->getFixtures(), append: true);

By default the ``ORMExecutor`` will wrap the purge and the load of fixtures
in a single transaction, which is the recommended way, but in some cases (for
example if loading your fixtures is too slow and causes timeouts) you may
want to wrap the purge and the load of every fixture in its own transaction.
To do so, you can use ``MultipleTransactionORMExecutor``.

.. code-block:: php
    <?php

>     $executor = new MultipleTransactionORMExecutor(new ORMExecutor($entityManager));

Reviewer: From a user perspective, it feels unnecessarily complicated that I have to nest both executors into each other.

Reviewer 2: There could be a static method that could act as a named constructor to avoid this.

.. code-block:: php
    <?php
    $executor->execute($loader->getFixtures(), append: true);

By default the ``ORMExecutor`` will wrap the purge and the load of fixtures
in a single transaction. You can pass ``false`` to the third argument

> to wrap the purge and the load of every fixture in its own transaction:

Reviewer 2: As a user, why would I want to do that?

Author: I updated the text with an example, I don't know if you had another example in mind but that was the issue for me.

Reviewer 2: > I don't know if you had another example in mind

I didn't have any reason in mind, I don't need that change myself. To be clear, your issue would be solved by disabling transactions entirely, but you'd rather have transactions on a fixture per fixture basis because… it's easier to start from there I suppose? The only advantage I can see is that will have fewer things to cleanup (you won't have to cleanup anything for the fixture that failed).

Author: I think it's good to still have transaction per fixture in case of failure, as you said, and yes, it's mostly because it's easier to have a flag to change this instead of disabling the transactions entirely (typically in a CI context).

Reviewer 2: So if there was a flag to disable transactions entirely, that would fix your issue as well, right? I wonder if using a boolean here is the right thing to do. I'm foreseeing issues where people have their issues fixed by disabling transactions entirely, but not by using a `per-fixture` mode.  Maybe we should have this instead:

```php
/* Convert this to enum when dropping support for PHP 7 */
interface TransactionMode
{
    public const SINGLE = 'single';
    public const PER_FIXTURE = 'per_fixture';
    public const DISABLED = 'disabled';
}
```

I'm not saying you are supposed to implement all modes, but I think using an enum-like will leave some room for additional modes if needed. Also, I think you would fix more issues by implementing `DISABLED` rather than `PER_FIXTURE`.

Author: That would also solve the issue yes, what I like about having one transaction per fixture is that it's often pretty close to the business logic in HTTP context, while a single transaction for everything will never happen.

.. code-block:: php
    <?php
    $executor->execute($loader->getFixtures(), append: true);

By default the ``ORMExecutor`` will wrap the purge and the load of fixtures
in a single transaction, which is the recommended way, but in some cases (for
example if loading your fixtures is too slow and causes timeouts) you may
want to wrap the purge and the load of every fixture in its own transaction.

> To do so, you can pass ``false`` to the third argument of ``ORMExecutor``:

Reviewer: This is a bit odd. You talk about a third argument, but the code uses named arguments. Maybe we should rephrase that.

```suggestion
To do so, construct ``ORMExecutor`` with ``singleTransaction: false``:
```

Author: I agree, I did it like this because of the previous sentence, shall I change it too?

> If you want to append the fixtures instead of purging before loading
> then pass ``true`` to the 2nd argument of execute:

Reviewer: If you'd like to, yes. 🙂

.. code-block:: php
    <?php

>     $executor->execute($loader->getFixtures(), singleTransaction: false);

Reviewer: This does not look right. `$singleTransaction` is a constructor parameter of `ORMExecutor`, not a parameter of the `execute()` method.

Author: My bad! Thanks for the notice
